torch/optim/lbfgs.py

```diff
@@ -239,3 +252,136 @@ def step(self, closure):
         state['prev_loss'] = prev_loss
 
         return orig_loss
+
+    def _copy_param(self):
+        original_param_data_list = []
+        for p in self._params:
+            param_data = p.data.new(p.size())
+            param_data.copy_(p.data)
+            original_param_data_list.append(param_data)
+        return original_param_data_list
+
+    def _set_param(self, param_data_list):
+        for i in range(len(param_data_list)):
+            self._params[i].data.copy_(param_data_list[i])
+
+    def _set_param_incremental(self, alpha, d):
+        offset = 0
+        for p in self._params:
+            numel = p.numel()
+            p.data.copy_(p.data + alpha*d[offset:offset + numel].resize_(p.size()))
+            offset += numel
+        assert offset == self._numel()
+
+    def _directional_derivative(self, d):
+        deriv = 0.0
+        offset = 0
+        for p in self._params:
+            numel = p.numel()
+            deriv += torch.sum(p.grad.data * d[offset:offset + numel].resize_(p.size()))
+            offset += numel
+        assert offset == self._numel()
+        return deriv
+
+    def _max_alpha(self, d):
+        offset = 0
+        max_alpha = float('inf')
+        for p, bnd in zip(self._params, self._bounds):
+            numel = p.numel()
+            l_bnd, u_bnd = bnd
+            p_grad = d[offset:offset + numel].resize_(p.size())
+            if l_bnd is not None:
+                from_l_bnd = ((l_bnd-p.data)/p_grad)[p_grad<0]
+                min_l_bnd = torch.min(from_l_bnd) if from_l_bnd.numel() > 0 else max_alpha
```

Comment:
This line takes the minimal alpha across all the directions. This would imply that if the point is near the boundary, the next step would be constrained to remain very close. I would expect instead to do different step size in each directions, see 16.71 mentioned in previous comment.